### PR TITLE
Adding basic features

### PR DESCRIPTION
**Pull request by Tojiddin20**
- Mostly refactored code, but also added logic for categories and users
- There are 2 types of categories, global and user
- Added AuthController, but we need to refactor and enhance it
- Added EmailService for sending registration emails
- Removed try catches, because they were ugly, we need to find a better solution for handling global errors

### CODE REVIEW

File: src/Context/ApplicationDbContext.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
+using Microsoft.EntityFrameworkCore;
+using server.Models;
+
+namespace server.Context;
+
+public class ApplicationDbContext : DbContext
+{
+    public DbSet<User> Users { get; set; } = null!;
```
Comment (veyxov) [inline src/Context/ApplicationDbContext.cs]: please use `public DbSet... set; } = Set<User>();`

Comment (veyxov) [reply src/Context/ApplicationDbContext.cs]: instead of `= null!;`

File: src/Controllers/AuthController.cs [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    {
+        User user = new User
+        {
+            Name = createUserDto.Name,
+            Email = createUserDto.Email,
+            PasswordHash = BCrypt.Net.BCrypt.HashPassword(createUserDto.Password),
+            EmailConfirmationToken = "Token" // todo 
+        };
+
+        UserValidator validator = new UserValidator();
```
Comment (veyxov) [inline src/Controllers/AuthController.cs]: Why are you validating the dto by hand? Create a middleware. I think fluent validation has a package for this: `FluentValidation.AspNetCore` something like that

File: src/Controllers/AuthController.cs [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        ValidationResult validationResult = await validator.ValidateAsync(user);
+        if (!validationResult.IsValid)
+        {
+            return ValidationProblem(validationResult.ToString());
+        }
+
+        SendEmailConfirmation(user);
```
Comment (veyxov) [inline src/Controllers/AuthController.cs]: This should be async

File: src/Controllers/AuthController.cs [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        await _dbContext.SaveChangesAsync();
+
+        return Ok(user);
+    }
+
+    [HttpPost("login")]
+    public async Task<ActionResult<User>> Login(LoginUserDto userDto)
+    {
+        User? user = await _dbContext.Users.FirstOrDefaultAsync(user => user.Email == userDto.Email);
+        if (user == null)
```
Comment (veyxov) [inline src/Controllers/AuthController.cs]:
```suggestion
        if (user is null)
```

File: src/Controllers/AuthController.cs [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+        {
+            return NotFound($"User with email {userDto.Email} not found!");
+        }
+
+        if (BCrypt.Net.BCrypt.EnhancedVerify(userDto.Password, user.PasswordHash))
```
Comment (veyxov) [inline src/Controllers/AuthController.cs]: Why `Enhanced`? Isn't it slower? Do we need this level of security?

File: src/Controllers/ExpenseController.cs [hunk 6]
```diff
@@ -1,137 +1,108 @@
...
+
+        await _dbContext.Expenses.AddAsync(expense);
+        await _dbContext.SaveChangesAsync();
+
+        return Ok(expense);
     }
 
     [HttpPut("{id}")]
-    public async Task<ActionResult<ExpenseDto>> Update(int id, [FromBody] ExpenseDto expenseDto)
+    public async Task<ActionResult<Expense>> Update(int id, UpdateExpenseDto updateExpenseDto)
```
Comment (veyxov) [inline src/Controllers/ExpenseController.cs]: Returning the main model is not good

File: src/Controllers/ExpenseController.cs [hunk 7]
```diff
@@ -1,137 +1,108 @@
...
-        catch (ValidationException exception)
+        Expense? expense = await _dbContext.Expenses.FindAsync(id);
+        if (expense is null)
         {
-            return ValidationProblem(exception.Message);
+            return NotFound($"Expense with id {id} not found!");
         }
-        catch (Exception exception)
+
+        expense.Amount = updateExpenseDto.Amount;
```
Comment (veyxov) [inline src/Controllers/ExpenseController.cs:50]: Use constructor or 
```
var something = new SomeType {
  a = dto.a,
  b = dto.b, 
};
```
or 
```
var model = dto.MapTo<Model>();
```

Comment (veyxov) [reply src/Controllers/ExpenseController.cs:50]: This is update, Ok

File: src/Program.cs [hunk 8]
```diff
@@ -1,10 +1,21 @@
...
+
 var builder = WebApplication.CreateBuilder(args);
 
 builder.Services.AddControllers();
 
 builder.Services.AddEndpointsApiExplorer();
 builder.Services.AddSwaggerGen();
 
+builder.Services.Configure<EmailSettings>(builder.Configuration.GetSection("EmailSettings"));
+builder.Services.AddTransient<IEmailService, EmailService>();
```
Comment (veyxov) [inline src/Program.cs]: Why transient?
t

File: src/Validators/CategoryValidator.cs [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+    public CategoryValidator(ApplicationDbContext dbContext)
+    {
+        _dbContext = dbContext;
+
+        RuleFor(category => category.Name)
+            .NotEmpty()
+            .Length(3, 20)
+            .WithMessage("Please provide a valid name with minimum length 3 and maximum 20")
+            .Must(BeAUniqueName)
```
Comment (veyxov) [inline src/Validators/CategoryValidator.cs]: Inserting business logic into the validator?

File: src/sqlite.db
Comment (veyxov) [inline src/sqlite.db:1]: Please remove database files for the PR

**Review by veyxov — CHANGES_REQUESTED** [review]

**Comment by veyxov** [comment]
>     * Removed try catches, because they were ugly, we need to find a better solution for handling global errors
You can use the `Result<T>` pattern for error handling.

File: src/Program.cs (hunk 8 above)
Comment (Tojiddin20) [reply src/Program.cs]: корочи 3 та бдай, транзиет, скопе и синглетоне. Хами созай барои и кора (памята холи мена)

File: src/Validators/CategoryValidator.cs (hunk 9 above)
Comment (Tojiddin20) [reply src/Validators/CategoryValidator.cs]: Корочи, Исмоил мега соз нестай истифода бурдани БД да валидатор. Чзе ки ба бизнес дахл дора вай бизнес лоҷикай

Comment (Tojiddin20) [reply src/Validators/CategoryValidator.cs]: Соответственно внутри валидации ты должен просто проверить данные на валидность, а то что связанно с логикой - отдельно

**Review by veyxov — APPROVED** [review]

**Pull request merged**